// teradata-nodejs-driver/teradata-logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeradataLogging = void 0;
class TeradataLogging {
    constructor(logLevel = 0) {
        this.traceLog = false;
        this.debugLog = false;
        this.dumpLog = false;
        // normally don't allow bitwise operations but are here
        /* tslint:disable-next-line */
        this.traceLog = (logLevel & 1) !== 0;
        /* tslint:disable-next-line */
        this.debugLog = (logLevel & 2) !== 0;
        /* tslint:disable-next-line */
        this.dumpLog = (logLevel & 4) !== 0;
    }
    infoLogMessage(message, jsonStringify) {
        if (jsonStringify) {
            message = JSON.stringify(message);
        }
        /* tslint:disable-next-line */
        console.log(message);
    }
    errorLogMessage(message, jsonStringify) {
        if (jsonStringify) {
            message = JSON.stringify(message);
        }
        /* tslint:disable-next-line */
        console.log(message);
    }
    traceLogMessage(message, jsonStringify) {
        if (this.traceLog) {
            if (jsonStringify) {
                message = JSON.stringify(message);
            }
            /* tslint:disable-next-line */
            console.log(message);
        }
    }
    debugLogMessage(message, jsonStringify) {
        if (this.debugLog) {
            if (jsonStringify) {
                message = JSON.stringify(message);
            }
            /* tslint:disable-next-line */
            console.log(message);
        }
    }
    dumpLogMessage(message, jsonStringify) {
        if (this.dumpLog) {
            if (jsonStringify) {
                message = JSON.stringify(message);
            }
            /* tslint:disable-next-line */
            console.log(message);
        }
    }
}
exports.TeradataLogging = TeradataLogging;
